src/content/articles/narrowing-function-parameters-with-rests-and-tuples.mdx

```diff
@@ -0,0 +1,315 @@
+
+const [fruit, info]: FruitAndInfo = Math.random()
+  ? ["apple", { color: "green" }]
+  : ["banana", { curvature: 35 }];
+
+if (fruit === "apple") {
+  info;
+  // ^?
+}
+```
+
+`logFruitTuple`'s `[fruit, info]` parameter is typed as `FruitAndInfo`, so it benefits from the same smartness around narrowing tuple types.
+
+## Which Should You Use?
+
+It depends.
+
+Most of the time, these kinds of complex function signatures are a sign of overly complicated code.
+Types that are difficult to represent in the type system are generally difficult because they're conceptually complex.
+Complexity is hard for developers to work with.
+When possible, try to avoid needing this kind of function signature.
+
+### Alternative: Objects
+
+Instead of wrestling with function parameters, consider using a single object to store the function's parameters.
+Doing so would allow using techniques such as the following _discriminated unions_ to represent different allowed shapes.
+
+```ts twoslash
+// @errors: 2322
+interface AppleInfo {
+  color: "green" | "red";
+}
+
+interface AppleAndInfo {
+  fruit: "apple";
+  info: AppleInfo;
+}
+
+interface BananaInfo {
+  curvature: number;
+}
+
+interface BananaAndInfo {
+  fruit: "banana";
+  info: BananaInfo;
+}
+
+function logFruitTuple({ fruit, info }: AppleAndInfo | BananaAndInfo) {
+  switch (fruit) {
+    case "apple":
+      console.log(`My apple's color is ${info.color}.`);
+      break;
+    case "banana":
+      console.log(`My banana's curvature is ${info.curvature}.`);
+      break;
+  }
+}
+
+logFruitTuple({ fruit: "apple", info: { color: "green" } }); // Ok
+
+logFruitTuple({ fruit: "banana", info: { color: "green" } }); // Should error
+```
+
+### Technique Tradeoffs
+
+If you must use one of these techniques, consider which one will cause your project's developer(s) the least pain in the future.
+Many TypeScript developers prefer to avoid function overloads because of how wacky they appear at first glance.
+Generics can also be intimidating to many TypeScript developers, though generally less so in simple cases such as this one.
+
+`...` rest tuple parameters are a nifty trick that also run the risk of confusing developers.
+But the increased type safety internally may be worth that confusion.
+
+Wrapping arguments in an array and then immediately destructuring that array also runs a small risk of hurting runtime performance if used in a very commonly run path.
+Though, this performance hit is likely optimized to be zero or near-zero in most instances of the function.
+Always diagnose runtime performance problems before jumping to conclusions.
+
+## Further Reading
+
+- Narrowing is covered in _Learning TypeScript_ Chapter 3: Unions and Literals.
+- Object types and discriminated type unions are covered in _Learning TypeScript_ Chapter 4: Objects.
+- Function overloads are covered in _Learning TypeScript_ Chapter 5: Functions.
+- Generic functions are covered in _Learning TypeScript_ Chapter 10: Generics.
```

Comment:
[suggestion]
Are you going to hyperlink these parts? That would be helpful for people who want to learn more.